### PR TITLE
improve regex overrides

### PR DESCRIPTION
**Pull request by ElenaDiachenko**
## Description

- Regex overrides improvements (prevent duplication)
- `npx rnv patch reset` => revert applied overrides

## Related issues

- https://github.com/flexn-io/renative/issues/1691

## Npm releases

n/a

## Testing 

- `yarn bootstrap-clean`
- run any platform => overrides were applied ((to make sure you can check node_modules by following the path in the summary)  and  .rnv/overrides folder was created in the root
-  when you run the same command again, overrides should not be applied a second time
- you can change any override and it should change in `.rnv/overrides/applied_overrides.json` and be applied to node_modules.
- run `npx rnv patch reset` => reverts applied overrides + deletes .rnv/overrides folder

### CODE REVIEW

**Review by locksten — CHANGES_REQUESTED** [review]
Adding marker comments like this:
```ts
import * as NativeComponentRegistry from '../../NativeComponent/NativeComponentRegistry';
import {ConditionallyIgnoredEventHandlers} from '../../NativeComponent/ViewConfigIgnore';
import Platform from '../../Utilities/Platform';/*RNV*/
import NativeModules from '../../BatchedBridge/NativeModules';/*RNV*/
```
would break overrides in files that don't support `/**/` comments, like `.podspec` or `.properties`, as well as overrides in multiline strings or macros.
But in most use cases it would be fine, so just restricting it to languages that support `/**/` comments might be enough, but it would make overrides work inconsistently.

File: packages/core/src/plugins/index.ts [hunk 1]
```diff
@@ -728,69 +730,166 @@ const _overridePlugin = (c: RnvContext, pluginsPath: string, dir: string) => {
...
+        const { backupPath, isFirstRun } = _saveOriginalFile(dest, overrideDir);
+        const packageName = _getPackageName(dest, fileKey);
+        const previousOverride = appliedOverrides[packageName]?.[fileKey] || {};
+        const overridesChanged = JSON.stringify(previousOverride) !== JSON.stringify(override);
+        if (overridesChanged && !isFirstRun) {
+            revertOverrideToOriginal(dest, backupPath);
+            fileToFix = fsReadFileSync(dest).toString();
+        }
         let foundRegEx = false;
+        const markerComment = '/*RNV*/';
```
Comment (locksten) [inline packages/core/src/plugins/index.ts:763]: Are the `/*RNV*/` marker comments still necessary now that overrides are applied to originals?

File: packages/core/src/plugins/index.ts [hunk 2]
```diff
@@ -728,69 +730,166 @@ const _overridePlugin = (c: RnvContext, pluginsPath: string, dir: string) => {
...
+    }
+};
+const _getPackageName = (dest: string, fileKey: string): string => {
+    const nodeModulesIndex = dest.lastIndexOf('node_modules');
+    if (nodeModulesIndex !== -1) {
+        const relativePath = dest.slice(nodeModulesIndex + 'node_modules'.length + 1);
+        const packageName = relativePath.replace(`/${fileKey}`, '');
+        return packageName;
     }
+    return '';
```
Comment (locksten) [inline packages/core/src/plugins/index.ts:828]: Why does this `return ''` while `_saveOriginalFile` throws an error?

File: packages/engine-core/src/tasks/patch/taskPatchReset.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        const isMonorepo = getConfigRootProp('isMonorepo');
+        const overrideDir = isMonorepo
+            ? path.join(ctx.paths.project.dir, '../../.rnv', 'overrides')
+            : path.join(ctx.paths.project.dir, '.rnv', 'overrides');
+
+        if (!fsExistsSync(overrideDir)) {
+            return logWarning(`Plugin overrides have not been applied yet`);
+        }
+        const rootPath = isMonorepo ? path.join(ctx.paths.project.dir, '../..') : path.join(ctx.paths.project.dir);
+        const appliedOverrideFilePath = path.join(overrideDir, 'applied_overrides.json');
```
Comment (locksten) [inline packages/engine-core/src/tasks/patch/taskPatchReset.ts:26]: Is it expected that `patch reset` does not revert file-based overrides, only `overrides.json`? If so, the description should specify that, and it should maybe print a warning.

File: packages/core/src/plugins/index.ts [hunk 4]
```diff
@@ -728,69 +730,166 @@ const _overridePlugin = (c: RnvContext, pluginsPath: string, dir: string) => {
...
 
-export const overrideFileContents = (dest: string, override: Record<string, string>, overridePath = '') => {
     if (fsExistsSync(dest)) {
         let fileToFix = fsReadFileSync(dest).toString();
-
+        const { backupPath, isFirstRun } = _saveOriginalFile(dest, overrideDir);
+        const packageName = _getPackageName(dest, fileKey);
+        const previousOverride = appliedOverrides[packageName]?.[fileKey] || {};
+        const overridesChanged = JSON.stringify(previousOverride) !== JSON.stringify(override);
+        if (overridesChanged && !isFirstRun) {
```
Comment (locksten) [inline packages/core/src/plugins/index.ts:758]: It seems like after reinstalling node_modules or updating a package, overrides don't get applied to the new clean node_modules files. And after changing an override, an outdated backup is restored and then overridden.

**Review by locksten — CHANGES_REQUESTED** [review]

**Review by locksten — COMMENTED** [review]
The `resetOverrides` hook won't automatically run in users' projects. What exactly is it's purpose? What's the difference between it and `taskPatchReset`?

---

It seems like overrides are also applied to subdependencies now:
```json
"@react-native-community/cli": {},
"react-native/node_modules/@react-native-community/cli": {},
"react-native-tvos/node_modules/@react-native-community/cli": {},
```
but weren't before (in b16b40ce0cb5458b3e61417fd5cb9c4759142c7c):
```json
"@react-native-community/cli": {},
```
@RicardasN  had an issue with overriding subdependencies a little while ago. Is this an intentional change for that?

**Review by locksten — APPROVED** [review]

**Review by pauliusguzas — APPROVED** [review]
just merge e2e file changes from main

**Pull request merged**